src/Servers/Kestrel/shared/KnownHeaders.cs

```diff
@@ -698,62 +727,95 @@ protected override bool AddValueFast(string key, StringValues value)
+                            _headers._raw{header.Identifier} = null;")}
+                            return true;
+                        }}
+                        return false;")}
+                    }}")}
+                    break;
+                }}")}
             }}
-{(loop.ClassName == "HttpResponseHeaders" ? @"
-            ValidateHeaderNameCharacters(key);" : "")}
-            Unknown.Add(key, value);
-            // Return true, above will throw and exit for false
-            return true;
+
+            return AddValueUnknown(key, value);
         }}
 
         protected override bool RemoveFast(string key)
         {{
             switch (key.Length)
             {{{Each(loop.HeadersByLength, byLength => $@"
                 case {byLength.Key}:
-                    {{{Each(byLength, header => $@"
-                        if (""{header.Name}"".Equals(key, StringComparison.OrdinalIgnoreCase))
-                        {{{(header.Identifier == "ContentLength" ? @"
-                            if (_contentLength.HasValue)
-                            {
-                                _contentLength = null;
-                                return true;
-                            }
-                            return false;" : $@"
-                            if ({header.TestBit()})
-                            {{
-                                {header.ClearBit()};
-                                _headers._{header.Identifier} = default(StringValues);{(header.EnhancedSetter == false ? "" : $@"
-                                _headers._raw{header.Identifier} = null;")}
-                                return true;
-                            }}
-                            return false;")}
-                        }}")}
-                    }}
-                    break;")}
+                {{{Each(byLength.OrderBy(h => !h.PrimaryHeader), header => $@"
+                    if (ReferenceEquals(HeaderNames.{header.Identifier}, key))
+                    {{{(header.Identifier == "ContentLength" ? @"
+                        if (_contentLength.HasValue)
+                        {
+                            _contentLength = null;
+                            return true;
+                        }
+                        return false;" : $@"
+                        if ({header.TestBit()})
+                        {{
+                            {header.ClearBit()};
+                            _headers._{header.Identifier} = default(StringValues);{(header.EnhancedSetter == false ? "" : $@"
+                            _headers._raw{header.Identifier} = null;")}
+                            return true;
+                        }}
+                        return false;")}
+                    }}")}
+    {Each(byLength.OrderBy(h => !h.PrimaryHeader), header => $@"
+                    if (HeaderNames.{header.Identifier}.Equals(key, StringComparison.OrdinalIgnoreCase))
```

Comment:
Everywhere there are two if blocks in a row with identical bodies and one checking object.ReferenceEquals and the other checking string.Equals, I have the same questions I did for TryGetValueFast.

Reply:
As above. Checks all the ReferenceEquals first; then does OrdinalIgnoreCase; rather than interleaving the IgnoreCase checks